### ISSUE TITLE
Packaged REST API client

### ISSUE THREAD

**Issue by virtualtam** [shaarli/Shaarli#745]
Hi!

I've started to write a simple REST client to test API-related Pull Requests (see early stubs on #739); would there be interest in providing an "official" Shaarli client to manipulate a Shaarli instance through its REST API?

The client would:
- use Python 3
- rely on Python built-ins and established libraries:
    - [Requests](http://docs.python-requests.org/en/master/)
    - [PyJWT](https://github.com/jpadilla/pyjwt)
    - [youtube-dl](https://github.com/rg3/youtube-dl/)
    - [PyAV](https://github.com/mikeboers/PyAV)
    - ... ;D
- be unit-tested with [Tox](https://tox.readthedocs.io/en/latest/) and [pytest](http://docs.pytest.org/en/latest/)
- be packaged and available from the [Python Package Index](https://pypi.python.org/pypi)
- feature a [Sphinx](http://www.sphinx-doc.org/en/1.5.1/) documentation hosted on [ReadTheDocs](https://readthedocs.org/)

And could potentially:
- be used as a library for users to develop custom clients and third-party integrations
- support legacy Shaarli instances, i.e. integrate features from [shaarchiver](https://github.com/nodiscc/shaarchiver)

ping @nodiscc @ArthurHoaro :)

**Comment by ArthurHoaro** [comment]
That could definitely be useful, if you're willing to work on it!

However I don't see what youtube-dl and PyAV have to do with this...

**Comment by virtualtam** [comment]
youtube-dl is already used by shaarchiver to download media from shaares' URLs, and PyAV provides bindings to ffmpeg/libav, e.g. to only process the audio part of a downloaded video (might be overkill ^^)

**Comment by ArthurHoaro** [comment]
Oh ok. Well, why not. But at first, having a real client which is able to use the API properly would be awesome. Additional features should come after that.

**Comment by nodiscc** [comment]
I would definitely use a proper command-line client for Shaarli, and am willing to rewrite shaarchiver around the API. It was primarily a first experiment with Python (but proved useful), so a full rewrite is in order.

What it could use is a client that is just able to:
 - get links by name, url, tags, id, date/timerange, private...
 - pass each query result's contents as command-line arguments, to custom scripts (eg `scripts/*.py`).

Scripts could fulfill the role shaarchiver used to have, and stay out of the reference API client code. Some [example scripts](https://gist.github.com/nodiscc/e42fcf15dacd4e57c0abdac31e9afbbe) of what could be done, separately. Again these should stay out of the reference client code, with their own youtube-dl/~~libav~~ffmpeg/... dependencies, but it's good if the client provides comfort/flexibility to achieve more complex work.

In short I'd be glad to help on a python+requests+jwt+pytest (been wanting to try pytest for a long time) client :p and write assorted "extractors" to extend it.

(If you have good resources to share about python-jwt/tox, please do)!

**Comment by virtualtam** [comment]
> pass each query result's contents as command-line arguments

This can be an issue, I've seen Bash shell buffers overflowed and data lost this way ^^

> What it could use is a client that is just able to:
> - get links by name, url, tags, id, date/timerange, private...
> - pass each query result's contents as command-line arguments, to custom scripts (eg scripts/*.py).

> Scripts could fulfill the role shaarchiver used to have

A bare API connector would have quite a limited use (consisting in simple API calls), the added value comes from providing extra/useful features such as dumping links, downloading media, scraping content, etc.

This can be structured in a simple way:
- core API package (as in "Python package") to interact with a Shaarli instance using either the REST API or legacy HTTP calls
- additional packages to provide extended features
- CLI scripts to perform operations in a user-friendly way

Note that it's possible to mark some features as being optional, in which case they're available only if the user has installed the appropriate dependencies.

> I would definitely use a proper command-line client for Shaarli, and am willing to rewrite shaarchiver around the API. It was primarily a first experiment with Python (but proved useful), so a full rewrite is in order.

Should we fork `shaarchiver` and refactor the existing scripts a bit for packaging, then start adding REST features? Or start a new project from scratch?

**Comment by nodiscc** [comment]
I'd rather start fresh with a new project (core API package). @virtualtam If you can lay down the basics of the architecture you described above (I fully agree) in a new repo of the Shaarli org that would be great.

**Comment by virtualtam** [comment]
Cool!
@ArthurHoaro @nodiscc do you have a naming preference for the new project? I was thinking something in the lines of `python-shaarli-client`, more inspired ideas are welcome ;-)

**Comment by virtualtam** [comment]
Closing as work has been started, see https://github.com/shaarli/python-shaarli-client/
